pkg/common/commands/resetoffset/controller/offsets.go

```diff
@@ -0,0 +1,230 @@
+// function used when reconciling offsets which facilitates stubbing in unit tests.
+var SaramaNewOffsetManagerFromClientFn SaramaNewOffsetManagerFromClientFnType = sarama.NewOffsetManagerFromClient
+
+// updateOffsetsError is the general error for any failure in updating offsets, exposed for test use.
+var updateOffsetsError = fmt.Errorf("failed to update Offsets")
+
+// reconcileOffsets updates the Offsets of all Partitions for the specified
+// Topic / ConsumerGroup to the Offset value corresponding to the specified
+// offsetTime (millis since epoch) and return OffsetMappings of the old/new
+// state.  An error will be returned and the Offsets will not be committed
+// if any problems occur.
+func (r *Reconciler) reconcileOffsets(ctx context.Context, topicName string, groupId string, offsetTime int64) ([]kafkav1alpha1.OffsetMapping, error) {
+
+	// Get The Logger From The Context & Enhance The With Parameters
+	logger := logging.FromContext(ctx).Desugar().With(
+		zap.String("Topic", topicName),
+		zap.String("Group", groupId),
+		zap.Int64("Time", offsetTime))
+
+	// Initialize A New Sarama Client
+	//
+	// ResetOffset is an infrequently used feature so there is no need for
+	// reuse, and there are also "broken-pipe" failures (non-recoverable)
+	// after periods of inactivity to deal with...
+	//   https://github.com/Shopify/sarama/issues/1162
+	//   https://github.com/Shopify/sarama/issues/866
+	saramaClient, err := SaramaNewClientFn(r.kafkaBrokers, r.saramaConfig)
+	defer safeCloseSaramaClient(logger, saramaClient)
+	if saramaClient == nil || err != nil {
+		logger.Error("Failed to create a new Sarama Client", zap.Error(err))
+		return nil, err
+	}
+
+	// Get The Partitions Of The Specified Kafka Topic
+	partitions, err := saramaClient.Partitions(topicName)
+	if err != nil {
+		logger.Error("Failed to determine Partitions for Topic", zap.Error(err))
+		return nil, err
+	}
+	logger.Info("Found Topic Partitions", zap.Any("Partitions", partitions))
+
+	// Create An OffsetManager For The Specified ConsumerGroup
+	offsetManager, err := SaramaNewOffsetManagerFromClientFn(groupId, saramaClient)
+	if offsetManager == nil || err != nil {
+		logger.Error("Failed to create OffsetManager for ConsumerGroup", zap.Error(err))
+		return nil, err
+	}
+
+	// Track Partition Specific Resources
+	partitionOffsetManagers := make([]sarama.PartitionOffsetManager, len(partitions))
+	offsetMappings := make([]kafkav1alpha1.OffsetMapping, len(partitions))
+
+	// Loop Over The Partitions - Updating Offsets & Tracking Results
+	updatesFailed := false
+	for index, partition := range partitions {
+		partitionOffsetManager, offsetMapping, updateErr := updateOffset(logger, saramaClient, offsetManager, topicName, partition, offsetTime)
+		partitionOffsetManagers[index] = partitionOffsetManager
+		if offsetMapping != nil {
+			offsetMappings[index] = *offsetMapping
+		}
+		if updateErr != nil {
+			updatesFailed = true
+		}
```

Comment:
isn't the `offsetMapping` only `nil`, where there was an error in the update?

Reply:
Yeah, this handling I think is part of the overall complexity problem. the new rewrite should be simpler.